File: AngularJS/src/org/angular2/entities/metadata/psi/Angular2MetadataDirectiveProperty.java
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.entities.metadata.psi;

import com.intellij.lang.javascript.psi.JSRecordType;
import com.intellij.lang.javascript.psi.JSType;
import com.intellij.openapi.util.NotNullComputable;
import com.intellij.psi.PsiElement;
import org.angular2.codeInsight.Angular2LibrariesHacks;
import org.angular2.entities.Angular2DirectiveProperty;
import org.angular2.entities.Angular2EntityUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;
import java.util.function.Supplier;

import static com.intellij.util.ObjectUtils.doIfNotNull;

public class Angular2MetadataDirectiveProperty implements Angular2DirectiveProperty {

  private final Supplier<JSRecordType.PropertySignature> mySignatureSupplier;
  private final NotNullComputable<? extends PsiElement> mySourceSupplier;
  private final String myName;

  Angular2MetadataDirectiveProperty(@NotNull Supplier<JSRecordType.PropertySignature> signatureSupplier,
                                    @NotNull NotNullComputable<? extends PsiElement> sourceSupplier,
                                    @NotNull String name) {
    this.mySignatureSupplier = signatureSupplier;
    this.mySourceSupplier = sourceSupplier;
    this.myName = name;
  }

  @Override
  public @NotNull String getName() {
    return myName;
  }

  @Override
  public @Nullable JSType getType() {
    return doIfNotNull(mySignatureSupplier.get(),
                       signature -> Angular2LibrariesHacks.hackQueryListTypeInNgForOf(signature.getJSType(), this));
  }

  @Override
  public boolean isVirtual() {
    return mySignatureSupplier.get() == null;
  }

  @Override
  public @NotNull PsiElement getSourceElement() {
    return Optional.ofNullable(mySignatureSupplier.get())
      .map(sig -> sig.getMemberSource().getSingleElement())
      .orElse(mySourceSupplier.compute());
  }

  @Override
  public String toString() {
    return Angular2EntityUtils.toString(this);
  }
}

File: AngularJS/src/org/angular2/entities/metadata/psi/Angular2MetadataDirectiveProperty.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.angular2.entities.metadata.psi

import com.intellij.javascript.web.js.WebJSTypesUtil.wrapWithUndefinedIfOptional
import com.intellij.javascript.webSymbols.apiStatus
import com.intellij.lang.javascript.psi.JSElementBase
import com.intellij.lang.javascript.psi.JSRecordType
import com.intellij.lang.javascript.psi.JSType
import com.intellij.lang.javascript.psi.ecma6.TypeScriptClass
import com.intellij.model.Pointer
import com.intellij.openapi.util.NullableLazyValue
import com.intellij.openapi.util.NullableLazyValue.lazyNullable
import com.intellij.psi.PsiElement
import com.intellij.refactoring.suggested.createSmartPointer
import com.intellij.util.asSafely
import com.intellij.webSymbols.WebSymbolApiStatus
import com.intellij.webSymbols.utils.coalesceApiStatus
import com.intellij.webSymbols.utils.coalesceWith
import org.angular2.codeInsight.Angular2LibrariesHacks
import org.angular2.entities.Angular2DirectiveProperty
import org.angular2.entities.Angular2EntityUtils
import java.util.*

class Angular2MetadataDirectiveProperty internal constructor(private val myOwner: Angular2MetadataClassBase<*>,
                                                             private val myFieldName: String,
                                                             override val name: String,
                                                             override val kind: String) : Angular2DirectiveProperty {

  private val mySignature: NullableLazyValue<JSRecordType.PropertySignature> =
    lazyNullable { myOwner.getPropertySignature(myFieldName) }

  override val rawJsType: JSType?
    get() = mySignature.value?.let { Angular2LibrariesHacks.hackQueryListTypeInNgForOf(it.jsType, this)?.wrapWithUndefinedIfOptional(it) }

  override val virtualProperty: Boolean
    get() = mySignature.value == null

  override val apiStatus: WebSymbolApiStatus
    get() = coalesceApiStatus(mySignature.value?.memberSource?.allSourceElements) { (it as? JSElementBase)?.apiStatus }
      .coalesceWith(myOwner.sourceElement.asSafely<JSElementBase>()?.apiStatus)

  override val sourceElement: PsiElement
    get() = mySignature.value?.memberSource?.singleElement ?: myOwner.sourceElement

  override val required: Boolean
    get() = false

  override val owner: TypeScriptClass?
    get() = myOwner.typeScriptClass

  override fun createPointer(): Pointer<Angular2MetadataDirectiveProperty> {
    val owner = myOwner.createSmartPointer()
    val name = this.name
    val fieldName = myFieldName
    val kind = this.kind
    return Pointer {
      owner.dereference()?.let { Angular2MetadataDirectiveProperty(it, fieldName, name, kind) }
    }
  }

  override fun toString(): String {
    return Angular2EntityUtils.toString(this)
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val property = other as Angular2MetadataDirectiveProperty?
    return myFieldName == property!!.myFieldName &&
           name == property.name &&
           kind == property.kind &&
           myOwner == property.myOwner
  }

  override fun hashCode(): Int {
    return Objects.hash(myFieldName, name, kind, myOwner)
  }
}
